klippy/extras/dockable_probe.py

```diff
@@ -0,0 +1,596 @@
+    def cmd_DETACH_PROBE(self, gcmd):
+        return_pos = self.toolhead.get_position()
+        self.detach_probe(return_pos)
+
+    #######################################################################
+    # Functions for calculating points and moving the toolhead
+    #######################################################################
+
+    # Move the toolhead to minimum safe distance aligned with angle
+    def _move_to_vector(self, angle):
+        x, y = self._get_point_on_vector(self.dock_position[:2],
+                                         angle,
+                                         self.approach_distance)
+        self.toolhead.manual_move([x,y,None], self.travel_speed)
+
+    # Move the toolhead to angle within minimium safe distance
+    def _align_to_vector(self, angle):
+        approach = self._get_intercept(self.toolhead.get_position(),
+                                       angle + (pi/2),
+                                       self.dock_position,
+                                       angle)
+        self.toolhead.manual_move([approach[0], approach[1], None],
+                                  self.attach_speed)
+
+    # Determine toolhead distance to dock coordinates
+    def _check_distance(self, pos=None, dist=None):
+        if not pos:
+            pos = self.toolhead.get_position()
+        dock = self.dock_position
+
+        if dist > sqrt((pos[0]-dock[0])**2 +
+                       (pos[1]-dock[1])**2 ):
+            return True
+        else:
+            return False
+
+    # Find a point on a vector line at a specific distance
+    def _get_point_on_vector(self, point, angle, magnitude=1):
+        x = point[0] - magnitude * cos(angle)
+        y = point[1] - magnitude * sin(angle)
+        return (x, y)
+
+    # Locate the intersection of two vectors
+    def _get_intercept(self, point1, angle1, point2, angle2):
+        x1, y1 = point1[:2]
+        x2, y2 = self._get_point_on_vector(point1, angle1, 10.0)
+        x3, y3 = point2[:2]
+        x4, y4 = self._get_point_on_vector(point2, angle2, 10.0)
+        det1 = ((x1 * y2) - (y1 * x2))
+        det2 = ((x3 * y4) - (y3 * x4))
+        d = ((x1 - x2) * (y3 - y4)) - ((y1 - y2) * (x3 - x4))
+        x = float((det1 * (x3 - x4)) - ((x1 - x2) * det2)) / d
+        y = float((det1 * (y3 - y4)) - ((y1 - y2) * det2)) / d
+        return (x, y)
+
+    # Determine the vector of two points
+    def _get_vector(self, point1, point2):
+        x1, y1 = point1[:2]
+        x2, y2 = point2[:2]
+        magnitude = sqrt((x2-x1)**2 + (y2-y1)**2 )
+        angle = atan2(y2-y1, x2-x1) + pi
+
+        return angle, magnitude
+
+    # Align z axis to prevent crashes
+    def _align_z(self):
+        if not self._last_homed:
+            curtime = self.printer.get_reactor().monotonic()
+            homed_axes = self.toolhead.get_status(curtime)['homed_axes']
+            self._last_homed = homed_axes
+        if self.dock_fixed_z:
```

Comment:
This `if` statement is a bit unwieldy, could maybe help to split as two functions e.g. `_align_z_fixed` and `align_z_moving` or similar which get called based on `self.dock_fixed_z`?

Reply:
I think that can be arranged.